File: WYPDFConverterOC/WYPDFConverter/WYPDFConverter.m
//
//  WYPDFConverter.m
//  WYPDFConverterOC
//
//  Created by Apple on 2017/6/20.
//  Copyright © 2017年 White-Young. All rights reserved.
//

#import "WYPDFConverter.h"

@implementation WYPDFConverter
+ (BOOL)convertPDFWithImages:(NSArray<UIImage *>*)images fileName:(NSString *)fileName{
    
    if (!images || images.count == 0) return NO;
    
    // pdf文件存储路径
    NSString *pdfPath = [self saveDirectory:fileName];
    NSLog(@"****************文件路径：%@*******************",pdfPath);

    BOOL result = UIGraphicsBeginPDFContextToFile(pdfPath, CGRectZero, NULL);
    
    // pdf每一页的尺寸大小
    
    CGRect pdfBounds = UIGraphicsGetPDFContextBounds();
    CGFloat pdfWidth = pdfBounds.size.width;
    CGFloat pdfHeight = pdfBounds.size.height;
    
    NSLog(@"%@",NSStringFromCGRect(pdfBounds));
    
    [images enumerateObjectsUsingBlock:^(UIImage * _Nonnull image, NSUInteger idx, BOOL * _Nonnull stop) {
        // 绘制PDF
        UIGraphicsBeginPDFPage();
        
        // 获取每张图片的实际长宽
        CGFloat imageW = image.size.width;
        CGFloat imageH = image.size.height;
//        CGRect imageBounds = CGRectMake(0, 0, imageW, imageH);
//        NSLog(@"%@",NSStringFromCGRect(imageBounds));
        
        // 每张图片居中显示
        // 如果图片宽高都小于PDF宽高
        if (imageW <= pdfWidth && imageH <= pdfHeight) {
         
            CGFloat originX = (pdfWidth - imageW) * 0.5;
            CGFloat originY = (pdfHeight - imageH) * 0.5;
            [image drawInRect:CGRectMake(originX, originY, imageW, imageH)];
            
        }
        else{
            CGFloat w,h; // 先声明缩放之后的宽高
//            图片宽高比大于PDF
            if ((imageW / imageH) > (pdfWidth / pdfHeight)){
                w = pdfWidth - 20;
                h = w * imageH / imageW;
                
            }else{
//             图片高宽比大于PDF
                h = pdfHeight - 20;
                w = h * imageW / imageH;
            }
            [image drawInRect:CGRectMake((pdfWidth - w) * 0.5, (pdfHeight - h) * 0.5, w, h)];
        }
    }];
    
    UIGraphicsEndPDFContext();
    
    return result;
}

+ (BOOL)convertPDFWithWebView:(UIWebView *)webView fileName:(NSString *)fileName{
    
    NSString *pdfPath = [self saveDirectory:fileName];
    NSLog(@"****************文件路径：%@*******************",pdfPath);

    NSData *pdfData = [webView convert2PDFData];
    BOOL result = [pdfData writeToFile:pdfPath atomically:YES];
    
    return result;
}


/**
 文件保存路径
 */
// 创建目录文件夹
+ (void)creatPDFFolder{
    NSFileManager *fileManager = [NSFileManager defaultManager];
    if (![fileManager fileExistsAtPath:[self pdfSaveFolder]]) {
        [fileManager createDirectoryAtPath:[self pdfSaveFolder] withIntermediateDirectories:YES attributes:nil error:NULL];
    }
}
//    文件存放主目录
+ (NSString *)pdfSaveFolder{
    return [[NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES) lastObject] stringByAppendingPathComponent:@"WYPDF"];
}

+ (NSString *)saveDirectory:(NSString *)fileName{
    [self creatPDFFolder];
    return [[self pdfSaveFolder] stringByAppendingPathComponent:fileName];
}

@end

File: WYPDFConverterSwift/WYPDFConverter/WYPDFConverter.swift
//
//  WYPDFConverter.swift
//  WYPDFConverterSwift
//
//  Created by Apple on 2017/6/21.
//  Copyright © 2017年 杨新威. All rights reserved.
//

import UIKit

class WYPDFConverter: NSObject {
    
   class func convertPDFWithImages(images : Array<UIImage>,fileName : String) -> Bool {
        return true
    }
    
   class func converterPDFWithWebView(_ webView : UIWebView, fileName : String) -> Bool {
        let pdfPath = self.saveDirectory(fileName : fileName)
        print("/******************文件类型****************/\n"+pdfPath+"\n/************************************/");
        
        let pdfData = self.convert2PDFData(webView: webView)
        
        let result = pdfData.write(toFile: self.saveDirectory(fileName: fileName), atomically: true)
        
        return result
    }
    
    fileprivate class func convert2PDFData(webView : UIWebView) ->NSData {
        
        
        let format = webView.viewPrintFormatter()
        let render = UIPrintPageRenderer.init()
        
        render.addPrintFormatter(format, startingAtPageAt: 0)
        
        let pageRect = CGRect.init(x: 0, y: 0, width: 600, height: 768)
        let printableRect = pageRect.insetBy(dx: 50, dy: 50)
        
        
        
        render.setValue(NSValue.init(cgRect: pageRect), forKey: "paperRect")
        render.setValue(NSValue.init(cgRect: printableRect), forKey: "printableRect")
        
        let pdfData = NSMutableData.init()
        
        UIGraphicsBeginPDFContextToData(pdfData, CGRect.zero, nil)
        
        for i in 0..<render.numberOfPages{
            UIGraphicsBeginPDFPage()
            render.drawPage(at: i, in: UIGraphicsGetPDFContextBounds())
        }
        
        UIGraphicsEndPDFContext()
        
        return pdfData
    }

    
   class func saveDirectory(fileName:String) -> String {
       
        let manager = FileManager.default
        let urlForDocument = manager.urls(for: .cachesDirectory, in: .userDomainMask);
        let url = urlForDocument[0]
        return url.path + "/" + fileName
    }
}
